src/coreclr/jit/emitxarch.cpp

```diff
@@ -34,8 +34,15 @@ bool emitter::IsSSEOrAVXInstruction(instruction ins)
     return (ins >= INS_FIRST_SSE_INSTRUCTION) && (ins <= INS_LAST_AVX_INSTRUCTION);
 }
 
+bool emitter::IsKInstruction(instruction ins)
+{
+    return (ins >= INS_FIRST_K_INSTRUCTION) && (ins <= INS_LAST_K_INSTRUCTION);
+}
+
 //------------------------------------------------------------------------
-// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse instruction.
+// IsAvx512OrPriorInstruction: Is this an Avx512 or Avx or Sse or K (opmask) instruction.
+// Technically, K instructions would be considered under the VEX encoding umbrella, but due to
+// the instruction table encoding had to be pulled out with the rest of the `INST5` definitions.
```

Comment:
Can you elaborate on this? Having two distinct ranges is going to make this check quite a bit more expensive.

Reply:
Looking, it looks like you're doing this to get access to `rr` which really exists to cover instructions like `inc`, `dec`, `push`, and `pop` which are "unary" and so are simply `inc reg` or `inc [addr]` -- It is definitely a confusing name and given that they encode the memory version using `mr`, we likely should've just used `rm` and removed the extra column from these...

Instructions like `kmovw` are instead `rm` or `mr` (`kmov reg, reg/mem` or `mov reg/mem, reg`) and so can be encoded using `INST4`

Reply:
I believe there's a restriction that the list of instructions in the table must be INST5, then INST4, then INST3, etc. The fact that we have ranges (e.g., FIRST_SSE_INSTRUCTION, LAST_SSE_INSTRUCTION) seems like a bit of luck? Instead of ranges should we use the flags bits instead?

Reply:
We would need to not use `INST5` then, since the way the tables are generated, they assume no gaps in the `INSTX` definitions, where `X` is a number. Correct?

Reply:
> I believe there's a restriction that the list of instructions in the table must be INST5, then INST4, then INST3, etc

Right. The point being that I believe we do not need (and actually shouldn't be using) INST5. We should be using INST3 for these instead and therefore these can be in the same general SIMD ranges so we got a simple check.

-- The reasoning being that these instructions are `ins mem, reg` (mr) or `ins reg, r/m` (rm) and you'll see that instructions like `mov` or `add` or SIMD instructions like `addps` or `movd` are encoded similarly. `rr` really just exists for the couple "unary" instructions which are simple `ins mem` or `ins reg`

Reply:
I understand what you mean, just want to make sure about this case for `kmov` though.

We have

`kmov mem, k reg`
`kmov k reg, k reg / mem
`kmov k reg, gp reg`
`kmov gp reg, k reg`

So the `mr` and `rm` forms do not handle when we want to move to and from general purpose registers.

Are you saying we can still use INST3?

Reply:
> Are you saying we can still use INST3?

Right

What we have is:
* `r, r/m` - `VEX.L0.0F.W0 90 /r KMOVW k1, k2/m16`
* `m, r  ` - `VEX.L0.0F.W0 91 /r KMOVW m16, k1`
* `r, r  ` - `VEX.L0.0F.W0 92 /r KMOVW k1, r32`
* `r, r  ` - `VEX.L0.0F.W0 93 /r KMOVW r32, k1`

For opcode `91`, we'll statically know it's correct based on it being the `emitIns_C_R`, `emitIns_S_R`, or similar memory encoding paths. So it can just be the `mr` entry.

For opcode `90`, we'll likewise statically know its correct based on it being `emitIns_R_C`, `emitIns_R_S`, or similar memory encoding paths. So it can just be the `rm` default entry.

As I indicated above, `rm` is also typically used for `reg, reg`. However, in this case we have a complexity in that we have three different opcodes that all map to the `reg, reg` scenario. This means that in `emitIns_R_R` we have to do some kind of check on both `isMaskReg(dstReg)` and `isMaskReg(op1Reg)`. Since we must do the checks to determine which of the 3 opcodes (`90`, `92`, or `93`) are correct, using `rr` doesn't benefit us at all and we're instead effectively best off defaulting to `90` and then changing it to `92` or `93` based on the other checks.